Mathlib/GroupTheory/SpecificGroups/Alternating/KleinFour.lean

```diff
@@ -0,0 +1,324 @@
+  rw [← lcm_cycleType] at hg
+  rw [mem_alternatingGroup, sign_of_cycleType] at hg0
+  have hg4 : g.cycleType.sum ≤ 4 := by
+    rw [← hα4, sum_cycleType, Nat.card_eq_fintype_card]
+    apply Finset.card_le_univ
+  by_cases h4 : 4 ∈ g.cycleType
+  · exfalso
+    suffices g.cycleType = {4} by
+      rw [this, ← Units.eq_iff] at hg0; norm_num at hg0
+    rw [← Multiset.cons_erase h4]
+    apply symm
+    rw [Multiset.singleton_eq_cons_iff]
+    apply And.intro rfl
+    rw [← Multiset.cons_erase h4] at hg4
+    simp only [Multiset.sum_cons, add_le_iff_nonpos_right,
+      le_zero_iff, Multiset.sum_eq_zero_iff] at hg4
+    ext x; simp only [Multiset.count_zero, Multiset.count_eq_zero]
+    intro hx
+    apply not_le.mpr (Equiv.Perm.one_lt_of_mem_cycleType (Multiset.mem_of_mem_erase hx))
+    simp [hg4 x hx]
+  · -- we know 4 ∉ g.cycleType,
+    suffices g.cycleType = Multiset.replicate (Multiset.card g.cycleType) 2 by
+      rw [this] at hg0
+      simp only [pow_add, Multiset.sum_replicate,
+        Multiset.card_replicate] at hg0
+      -- prove : Multiset.card g.cycleType ≤ 2
+      have hk2 : Multiset.card g.cycleType ≤ 2 := by
+        rw [this, Multiset.sum_replicate] at hg4
+        apply Nat.le_of_mul_le_mul_left
+        · rw [Nat.mul_comm 2]
+          exact hg4
+        norm_num
+      rcases Nat.eq_or_lt_of_le hk2 with hk2 | hk1
+      · -- g.cycleType.card = 2
+        rw [hk2] at this
+        simp only [this, Multiset.replicate_succ, Multiset.empty_eq_zero,
+          Multiset.cons_ne_zero, Multiset.insert_eq_cons, false_or,
+          Multiset.replicate_zero, Multiset.cons_zero]
+      · -- we know : g.cycleType.card ≤ 1
+        rw [Nat.lt_succ_iff] at hk1
+        rcases Nat.eq_or_lt_of_le hk1 with hk1 | hk0
+        · -- g.cycleType.card = 1 : exfalso
+          exfalso
+          simp [hk1] at hg0
+        · -- g.cycleType.card = 0
+          simp only [Nat.lt_one_iff] at hk0
+          left
+          simpa [hk0] using this
+    rw [Multiset.eq_replicate_card]
+    intro i hi
+    have : i ∣ 2 ^ n := by
+      apply Nat.dvd_trans _ hg
+      exact Multiset.dvd_lcm hi
+    rw [Nat.dvd_prime_pow Nat.prime_two] at this
+    obtain ⟨k, hk, hlcm⟩ := this
+    suffices k = 1 by simp [hlcm, this]
+    apply le_antisymm
+    · rw [← not_lt]
+      intro hk1
+      suffices k = 2 by
+        simp only [this, Nat.reducePow] at hlcm
+        apply h4
+        rwa [← hlcm]
+      refine le_antisymm ?_ hk1
+      · -- k ≤ 2
+        rw [← Nat.pow_le_pow_iff_right (Nat.le_refl 2)]
+        norm_num
+        rw [← hα4, ← hlcm, Nat.card_eq_fintype_card]
+        apply le_trans (le_card_support_of_mem_cycleType hi)
+        apply Finset.card_le_univ
+    · rw [Nat.one_le_iff_ne_zero]
+      rintro ⟨rfl⟩
+      rw [hlcm, pow_zero] at hi
+      apply Nat.lt_irrefl 1
+      exact one_lt_of_mem_cycleType hi
+
+theorem card_of_four (hα4 : Nat.card α = 4) :
+    Nat.card (alternatingGroup α) = 12 := by
```

Comment:
Can you add a general lemma `Nat.card_alternatingGroup` or something? (assuming ` [Nontrivial α]`)

Reply:
I had been reluctant of giving that with `(Nat.card alpha)  ! / 2 `, because except for explicit computations, that would require to know that `Nat.card alpha !` is even to be useful, but why not.